From 01828e4beaf1dd64cc02e487f06a0ae54e18394c Mon Sep 17 00:00:00 2001
From: JooJooBee666 <desire4homicide@gmail.com>
Date: Thu, 30 Jul 2015 17:46:01 -0400
Subject: [PATCH] libexfat: prevent segfault in exfat from NULL node in exfat_flush_node

Exfat/fuse/main.c fuse_exfat_fsync calls get_node(fi).  This returns a NULL
due to fi->fh (file handle) being NULL in certain instances.
exfat_flush_node does not check for a NULL before looking for node->parent.
This causes exfat to SEGFAULT. Unable to track down the root cause of
the NULL file handle being sent to fuse_exfat_release in fi. This issue
only occurs when using a 128GB SD Card and not a 32GB. No 64GB to test.
This is consistently triggered by a few applications.

Change-Id: I838941863b571f7357f3f48367a43c3f1d136a7d
---

diff --git a/libexfat/node.c b/libexfat/node.c
index 85d0726..8fca8e7 100644
--- a/libexfat/node.c
+++ b/libexfat/node.c
@@ -561,6 +561,12 @@
 	struct exfat_entry_meta1 meta1;
 	struct exfat_entry_meta2 meta2;
 
+	if (!node)
+	{
+		exfat_error("unable to flush node due to NULL node");
+		return 0; /* cannot flush without node info */
+	}
+
 	if (!(node->flags & EXFAT_ATTRIB_DIRTY))
 		return 0; /* no need to flush */
 
